src/Component/Carousel.php

```diff
@@ -0,0 +1,173 @@
+	 */
+	const FALLBACK_HEIGHT = 480;
+
+	/**
+	 * An object representation of the DOM.
+	 *
+	 * @var DOMDocument
+	 */
+	protected $dom;
+
+	/**
+	 * The images to add to the carousel.
+	 *
+	 * @var DOMElementList
+	 */
+	protected $images;
+
+	/**
+	 * Instantiates the class.
+	 *
+	 * @param DOMDocument    $dom    The dom to use to create a carousel.
+	 * @param DOMElementList $images The images from which to create a carousel.
+	 */
+	public function __construct( $dom, DOMElementList $images ) {
+		$this->dom    = $dom;
+		$this->images = $images;
+	}
+
+	/**
+	 * Gets the carousel element.
+	 *
+	 * @return DOMElement An <amp-carousel> with the images.
+	 */
+	public function get_dom_element() {
+		list( $width, $height ) = $this->get_dimensions();
+		$amp_carousel           = AMP_DOM_Utils::create_node(
+			$this->dom,
+			'amp-carousel',
+			[
+				'width'  => $width,
+				'height' => $height,
+				'type'   => 'slides',
+				'layout' => 'responsive',
+			]
+		);
+
+		foreach ( $this->images as $image ) {
+			$image_node = $image instanceof HasCaption ? $image->get_slide_node() : $image;
+			$caption    = $image instanceof HasCaption ? $image->get_caption() : null;
+			$slide      = AMP_DOM_Utils::create_node(
+				$this->dom,
+				'div',
+				[ 'class' => 'slide' ]
+			);
+
+			// Ensure the image fills the entire <amp-carousel>, so the possible caption looks right.
+			if ( 'amp-img' === $image_node->tagName ) {
+				$image_node->setAttribute( 'layout', 'fill' );
+				$image_node->setAttribute( 'object-fit', 'cover' );
+			} elseif ( isset( $image_node->firstChild->tagName ) && 'amp-img' === $image_node->firstChild->tagName ) {
+				// If the <amp-img> is wrapped in an <a>.
+				$image_node->firstChild->setAttribute( 'layout', 'fill' );
+				$image_node->firstChild->setAttribute( 'object-fit', 'cover' );
+			}
+
+			$slide->appendChild( $image_node );
+
+			// If there's a caption, wrap it in a <div> and <span>, and append it to the slide.
+			if ( $caption ) {
+				$caption_wrapper = AMP_DOM_Utils::create_node(
+					$this->dom,
+					'div',
+					[ 'class' => 'amp-wp-gallery-caption' ]
+				);
+				$caption_span    = AMP_DOM_Utils::create_node( $this->dom, 'span', [] );
+				$text_node       = $this->dom->createTextNode( $caption );
+
+				$caption_span->appendChild( $text_node );
+				$caption_wrapper->appendChild( $caption_span );
+				$slide->appendChild( $caption_wrapper );
+			}
+
+			$amp_carousel->appendChild( $slide );
+		}
+
+		return $amp_carousel;
+	}
+
+	/**
+	 * Gets the carousel's width and height, based on its images.
+	 *
+	 * This will return the width and height of the image with the widest aspect ratio,
+	 * not necessarily the image with the biggest absolute width.
+	 *
+	 * @return array {
+	 *     The carousel dimensions.
+	 *
+	 *     @type int $width  The width of the carousel, at index 0.
+	 *     @type int $height The height of the carousel, at index 1.
+	 * }
+	 */
+	public function get_dimensions() {
```

Comment:
Is this supposed to be publicly accessible, i.e. do we want to give access to this to third-party developers as our public interface?

Reply:
Uf, there's probably not a need for it to be public.

Reply:
https://github.com/ampproject/amp-wp/pull/3659/commits/478f71e28c4c333c31f1d4abc1c057af0379c0eb makes it `private`